Benchmarks/BuildingSimulation/Haskell/app/Main.hs

```diff
@@ -0,0 +1,318 @@
+                             , flow    = quanta.flow
+                             , density = quanta.density
+                             , cp      = quanta.cp
+                             }
+
+      tankMass     = store.volume * store.density
+      dTempTank    = (updatedPower * dTime) / store.cp / tankMass
+      updatedStore = Tank { temp    = store.temp + dTempTank
+                          , volume  = store.volume
+                          , cp      = store.cp
+                          , density = store.density
+                          , mass    = store.mass
+                          }
+  in
+  TankAndQuanta updatedStore updatedQuanta
+
+
+{-# SPECIALIZE lossCalc :: Float -> Float -> Float #-}
+{-# SPECIALIZE lossCalc :: Double -> Double -> Double #-}
+lossCalc :: Num a => a -> a -> a
+lossCalc pred gt =
+  let diff = pred - gt
+   in abs diff
+
+
+{-# SPECIALIZE simulate :: SimParams Float -> Float #-}
+{-# SPECIALIZE simulate :: SimParams Double -> Double #-}
+simulate :: Floating a => SimParams a -> a
+simulate (SimParams pexTube slab0 quanta0 tank0 temp0) =
+  let
+      slab0' = Slab temp0 slab0.area slab0.cp slab0.density slab0.thickness
+
+      go !i !slab !tank !quanta
+        | i >= timesteps = slab.temp
+        | otherwise      =
+            let TankAndQuanta tank' quanta'             = updateSourceTank tank quanta
+                QuantaAndPower quanta'' powerToBuilding = computeLoadPower slab pexTube (updateQuanta quanta')
+                slab'                                   = updateBuildingModel powerToBuilding slab
+            in
+            go (i+1) slab' tank' (updateQuanta quanta'')
+  in
+  go 0 slab0' tank0 quanta0
+
+
+fullPipe :: Floating a => SimParams a -> a
+fullPipe params =
+  let pred = simulate params
+      loss = lossCalc pred 27.344767
+  in
+  loss
+
+
+
+-- TLM: could probably at least make this a vector? The pack operation becomes
+-- more tedious however...
+unpack :: SimParams a -> [a]
+unpack (SimParams tube slab quanta tank startingTemp) =
+  toList tube <> toList slab <> toList quanta <> toList tank <> [startingTemp]
+
+pack :: [a] -> SimParams a
+pack x0 =
+  let (tube, x1)     = splitAt 4 x0
+      (slab, x2)     = splitAt 5 x1
+      (quanta, x3)   = splitAt 5 x2
+      (tank, x4)     = splitAt 5 x3
+      [startingTemp] = x4
+  in
+  SimParams (fromList tube) (fromList slab) (fromList quanta) (fromList tank) startingTemp
+      
+
+-- Simulation Parameters -------------------------------------------------------
+
+dTime :: Fractional a => a
+dTime = 0.1
+
+timesteps :: Int
+timesteps = 1000
+
+printGradToCompare :: Bool
+printGradToCompare = True
+
+main :: IO ()
+main = do
+  let params    = init :: SimParams Double
+      params'   = unpack params
+      fullPipe' = grad (fullPipe . pack)
+
+  when printGradToCompare $
+    print (pack (fullPipe' params'))
+
+  defaultMain
+    [ bench "primal"  $ nf fullPipe  params
+    , bench "adjoint" $ nf fullPipe' params'
```

Comment:
I'm guessing this would require installing `bench` locally? I do like the extra info from `bench`, but it would be nice to have each benchmark produce the same output -- i.e. https://github.com/PassiveLogic/differentiable-swift-examples/blob/main/Benchmarks/BuildingSimulation/Swift/main.swift#L248-L251

Reply:
Yes that `bench` function comes from the [criterion](https://hackage.haskell.org/package/criterion) package, which is doing the heavy lifting here.

Benchmarking in Haskell is a bit of a minefield with its lazy evaluation, but I could switch it to self-timed. Anybody who knows anything about Haskell will immediately ask "why didn't you use criterion" though.

![Screenshot 2024-10-31 at 09 19 04](https://github.com/user-attachments/assets/cc27b920-075c-435c-a60c-bc4c808cb869)

Reply:
I see. That report is pretty nice! The main thing I'm looking for is a way to directly compare with the other benchmarks, and continuity between them all. Does primal/adjoint map exactly to the forward/gradient measurements we record in the others? If they are, could we pull those out and print in addition to the `bench` output. If they're different, we may need to calculate separately.

Reply:
Yes I believe they are directly comparable; "primal" is evaluating the function normally (i.e. forward pass) and "adjoint" is computing the gradient in reverse-mode AD. So the "mean" number the report produces (rather than "time", which is the least-squares regression estimate) is more or less what the other benchmarks are computing (modulo it will pick the number of iterations to perform so that it has a good estimate, rather than fixing that to 1000).

Reply:
Updated so that it only prints a summary similar to the other benchmark programs:

```
average forward only time: 0.000025363051696440153 seconds
average forward and backwards (gradient) time: 0.004855364004974598 seconds
```